scripts/generate_messages.js

```diff
@@ -14,6 +14,16 @@
 
 'use strict';
 
+const path = require('path');
+
+let rootDir = path.dirname(__dirname);
+let rclnodejsTestMsgsPath = path.join(rootDir, 'test', 'rclnodejs_test_msgs');
+let actionPath = path.join(rootDir, 'test', 'ros1_actions');
+process.env.AMENT_PREFIX_PATH = process.env.AMENT_PREFIX_PATH + 
+  path.delimiter + rclnodejsTestMsgsPath +
+  path.delimiter + actionPath;
+console.log('AMENT_PREFIX_PATH:', process.env.AMENT_PREFIX_PATH);
+
 const generator = require('../rosidl_gen/index.js');
 
 console.log('Start to generate the JavaScript messages...');
```

Comment:
Please do not change this file, as `npm install` will execute it during installation. We'd better not change the `process.env.AMENT_PREFIX_PATH` variable, the paths added here are only for testing.

I guess you also have to make change to the `run_test.js`

Reply:
@minggangw , the main consideration of changing the `AMENT_PREFIX_PATH` is to generate js files for package `rclnodejs_test_msgs` and `ros1_actions` before testing. If this change must be removed from `npm install` then I have to generate the js packages twice. To make the test work, it's better to move the change to `compile_app.js`.

Reply:
For the current implement, all the JavaScript messages generated from IDL files are going to be removed before the running of test, please check [here](https://github.com/RobotWebTools/rclnodejs/blob/develop/scripts/run_test.js#L25)

``` JavaScript
fs.remove(path.join(path.dirname(__dirname), 'generated'), (err)=> {
  if (!err) {
    let mocha = new Mocha();
const testDir = path.join(__dirname, '../test/');
```

So I think you could append the path to `AMENT_PREFIX_PATH` in line [23](https://github.com/RobotWebTools/rclnodejs/blob/develop/scripts/run_test.js#L23). The messages will be re-generated before the test begins.

Reply:
I noticed that. However, the `ros1_action` package is a little tricky: it needs two rounds. 1st round will generate JavaScript message files and `CMakeLists.txt`. 2nd round will build the native libraries based on the `CMakeLists.txt`.  Maybe I should submit the generated `CMakeLists.txt` just like `rclnodejs_test_msgs`?

Reply:
I see and I suggest you could have the following steps:

* rclnodejs.regenerateAll() to generate the JS messages compulsively (a bunch of `.msg` and `CMakeLists.txt` will be present).
* Compile the `ros1_action` project by `CMakeList.txt` generated in step 1 and copy the installation files to the path of ROS 2.
*  Run mocha

Would you please investigate it to check if it's feasible? Thanks!

Reply:
Sure! I think it should work and I'll try it.